File: client/protocol/schemas.py
"""消息模式定义和验证"""
from typing import Any, Optional
from pydantic import BaseModel, Field, field_validator
from .message_types import MessageType
from .error_codes import ErrorCode
from .states import SessionState, DeviceStatus


class BaseMessage(BaseModel):
    """基础消息模型"""

    type: MessageType
    message_id: str = Field(..., min_length=1, max_length=64)
    timestamp: float = Field(..., gt=0)


class AuthMessage(BaseMessage):
    """认证消息"""

    type: MessageType = MessageType.AUTH
    token: str = Field(..., min_length=1)


class AuthSuccessMessage(BaseMessage):
    """认证成功消息"""

    type: MessageType = MessageType.AUTH_SUCCESS
    device_id: Optional[str] = None
    user_id: Optional[str] = None


class AuthFailureMessage(BaseMessage):
    """认证失败消息"""

    type: MessageType = MessageType.AUTH_FAILURE
    error_code: ErrorCode
    error_message: str


class SessionRequestMessage(BaseMessage):
    """会话请求消息"""

    type: MessageType = MessageType.SESSION_REQUEST
    session_id: str = Field(..., min_length=1, max_length=64)
    device_id: str = Field(..., min_length=1, max_length=64)
    operator_id: str = Field(..., min_length=1, max_length=64)


class SessionAcceptMessage(BaseMessage):
    """会话接受消息"""

    type: MessageType = MessageType.SESSION_ACCEPT
    session_id: str = Field(..., min_length=1, max_length=64)


class SessionRejectMessage(BaseMessage):
    """会话拒绝消息"""

    type: MessageType = MessageType.SESSION_REJECT
    session_id: str = Field(..., min_length=1, max_length=64)
    reason: str


class SessionEndMessage(BaseMessage):
    """会话结束消息"""

    type: MessageType = MessageType.SESSION_END
    session_id: str = Field(..., min_length=1, max_length=64)
    reason: Optional[str] = None


class SDPOfferMessage(BaseMessage):
    """SDP Offer 消息"""

    type: MessageType = MessageType.SDP_OFFER
    session_id: str = Field(..., min_length=1, max_length=64)
    sdp: str = Field(..., min_length=1)


class SDPAnswerMessage(BaseMessage):
    """SDP Answer 消息"""

    type: MessageType = MessageType.SDP_ANSWER
    session_id: str = Field(..., min_length=1, max_length=64)
    sdp: str = Field(..., min_length=1)


class ICECandidateMessage(BaseMessage):
    """ICE 候选消息"""

    type: MessageType = MessageType.ICE_CANDIDATE
    session_id: str = Field(..., min_length=1, max_length=64)
    candidate: str
    sdp_mid: Optional[str] = None
    sdp_m_line_index: Optional[int] = None


class HeartbeatMessage(BaseMessage):
    """心跳消息"""

    type: MessageType = MessageType.HEARTBEAT
    device_id: Optional[str] = None


class HeartbeatAckMessage(BaseMessage):
    """心跳确认消息"""

    type: MessageType = MessageType.HEARTBEAT_ACK


class StatusUpdateMessage(BaseMessage):
    """状态更新消息"""

    type: MessageType = MessageType.STATUS_UPDATE
    device_id: str = Field(..., min_length=1, max_length=64)
    status: DeviceStatus
    session_state: SessionState


class MetricsUpdateMessage(BaseMessage):
    """性能指标更新消息"""

    type: MessageType = MessageType.METRICS_UPDATE
    session_id: str = Field(..., min_length=1, max_length=64)
    fps: float = Field(..., ge=0, le=120)
    bitrate: int = Field(..., ge=0)
    rtt: float = Field(..., ge=0)
    packet_loss: float = Field(..., ge=0, le=1)
    cpu_usage: float = Field(..., ge=0, le=1)


class ErrorMessage(BaseMessage):
    """错误消息"""

    type: MessageType = MessageType.ERROR
    error_code: ErrorCode
    error_message: str
    details: Optional[dict[str, Any]] = None


class AckMessage(BaseMessage):
    """确认消息"""

    type: MessageType = MessageType.ACK
    ack_message_id: str = Field(..., min_length=1, max_length=64)


def validate_message(data: dict[str, Any]) -> BaseMessage:
    """验证并解析消息"""
    msg_type = data.get("type")
    if not msg_type:
        raise ValueError("Missing message type")

    message_classes = {
        MessageType.AUTH: AuthMessage,
        MessageType.AUTH_SUCCESS: AuthSuccessMessage,
        MessageType.AUTH_FAILURE: AuthFailureMessage,
        MessageType.SESSION_REQUEST: SessionRequestMessage,
        MessageType.SESSION_ACCEPT: SessionAcceptMessage,
        MessageType.SESSION_REJECT: SessionRejectMessage,
        MessageType.SESSION_END: SessionEndMessage,
        MessageType.SDP_OFFER: SDPOfferMessage,
        MessageType.SDP_ANSWER: SDPAnswerMessage,
        MessageType.ICE_CANDIDATE: ICECandidateMessage,
        MessageType.HEARTBEAT: HeartbeatMessage,
        MessageType.HEARTBEAT_ACK: HeartbeatAckMessage,
        MessageType.STATUS_UPDATE: StatusUpdateMessage,
        MessageType.METRICS_UPDATE: MetricsUpdateMessage,
        MessageType.ERROR: ErrorMessage,
        MessageType.ACK: AckMessage,
    }

    message_class = message_classes.get(MessageType(msg_type))
    if not message_class:
        raise ValueError(f"Unknown message type: {msg_type}")

    return message_class.model_validate(data)

File: frontend/src/protocol/schemas.ts
/**
 * 消息模式定义和验证
 */
import { MessageType } from "./messageTypes";
import { ErrorCode } from "./errorCodes";
import { SessionState, DeviceStatus } from "./states";

export interface BaseMessage {
  type: MessageType;
  message_id: string;
  timestamp: number;
}

export interface AuthMessage extends BaseMessage {
  type: MessageType.AUTH;
  token: string;
}

export interface AuthSuccessMessage extends BaseMessage {
  type: MessageType.AUTH_SUCCESS;
  device_id?: string;
  user_id?: string;
}

export interface AuthFailureMessage extends BaseMessage {
  type: MessageType.AUTH_FAILURE;
  error_code: ErrorCode;
  error_message: string;
}

export interface SessionRequestMessage extends BaseMessage {
  type: MessageType.SESSION_REQUEST;
  session_id: string;
  device_id: string;
  operator_id: string;
}

export interface SessionAcceptMessage extends BaseMessage {
  type: MessageType.SESSION_ACCEPT;
  session_id: string;
}

export interface SessionRejectMessage extends BaseMessage {
  type: MessageType.SESSION_REJECT;
  session_id: string;
  reason: string;
}

export interface SessionEndMessage extends BaseMessage {
  type: MessageType.SESSION_END;
  session_id: string;
  reason?: string;
}

export interface SDPOfferMessage extends BaseMessage {
  type: MessageType.SDP_OFFER;
  session_id: string;
  sdp: string;
}

export interface SDPAnswerMessage extends BaseMessage {
  type: MessageType.SDP_ANSWER;
  session_id: string;
  sdp: string;
}

export interface ICECandidateMessage extends BaseMessage {
  type: MessageType.ICE_CANDIDATE;
  session_id: string;
  candidate: string;
  sdp_mid?: string;
  sdp_m_line_index?: number;
}

export interface HeartbeatMessage extends BaseMessage {
  type: MessageType.HEARTBEAT;
  device_id?: string;
}

export interface HeartbeatAckMessage extends BaseMessage {
  type: MessageType.HEARTBEAT_ACK;
}

export interface StatusUpdateMessage extends BaseMessage {
  type: MessageType.STATUS_UPDATE;
  device_id: string;
  status: DeviceStatus;
  session_state: SessionState;
}

export interface MetricsUpdateMessage extends BaseMessage {
  type: MessageType.METRICS_UPDATE;
  session_id: string;
  fps: number;
  bitrate: number;
  rtt: number;
  packet_loss: number;
  cpu_usage: number;
}

export interface ErrorMessage extends BaseMessage {
  type: MessageType.ERROR;
  error_code: ErrorCode;
  error_message: string;
  details?: Record<string, any>;
}

export interface AckMessage extends BaseMessage {
  type: MessageType.ACK;
  ack_message_id: string;
}

export type SignalingMessage =
  | AuthMessage
  | AuthSuccessMessage
  | AuthFailureMessage
  | SessionRequestMessage
  | SessionAcceptMessage
  | SessionRejectMessage
  | SessionEndMessage
  | SDPOfferMessage
  | SDPAnswerMessage
  | ICECandidateMessage
  | HeartbeatMessage
  | HeartbeatAckMessage
  | StatusUpdateMessage
  | MetricsUpdateMessage
  | ErrorMessage
  | AckMessage;

export function validateMessage(data: any): SignalingMessage {
  if (!data.type) {
    throw new Error("Missing message type");
  }
  if (!data.message_id) {
    throw new Error("Missing message_id");
  }
  if (!data.timestamp || data.timestamp <= 0) {
    throw new Error("Invalid timestamp");
  }
  return data as SignalingMessage;
}
